### ISSUE TITLE
Rational Resampler is dropping stream tags

### ISSUE THREAD

**Issue by eltos** [gnuradio/gnuradio#7328]
### What happened?

The tag propagation for Rational Resampler Block seams to be broken: It sometimes undeterministically drops tags.
As the block is using the default tag propagation mechanism of the scheduler around [these lines](https://github.com/gnuradio/gnuradio/blame/72e21b54c44c360d511371420e74be4a6b2bfda6/gnuradio-runtime/lib/block_executor.cc#L159), the issue likely affects other interpolating blocks likewiese.
A Repeat Block used in the same place does not suffer from this issue (as it is handling tag propagation internally, and is not using the default tag propagation policy).

Possibly related to https://github.com/gnuradio/gnuradio/issues/2056, however here I am not changing interpolation at runtime, but using a fixed value from the beginning.

This happens rather rarely, and it is not yet clear what exactly causes the issue.

### System Information

OS: Linux Ubuntu
GR Installation Method: build from source

### GNU Radio Version

3.10 (maint-3.10)

### Specific Version

3.10.8.0

### Steps to Reproduce the Problem

Example flowgraph: [test_resampler_tag_propagation.zip](https://github.com/gnuradio/gnuradio/files/15252236/test_resampler_tag_propagation.zip)

![test_resampler_tag_propagation](https://github.com/gnuradio/gnuradio/assets/19860638/97ff85a7-0dd0-4430-9de8-4ed7b2f1085b)

Wait for a minute (or 2) until you spot this tag which got missing in the upper branch:

![rational resampler issue](https://github.com/gnuradio/gnuradio/assets/19860638/50ae068d-3550-475e-b9ae-ba72e0c918ac)

**Comment by ffu** [comment]
There should be a throttle block in the flow graph. If not, the cpu will run close to full speed, and I'm not willing to test that on my computer. With the throttle block, I cannot reproduce this error.

**Comment by eltos** [comment]
Well, throttling will just make the bug much more rare, so you'll have to wait proportionally longer until it eventually happens.

**Comment by marcusmueller** [comment]
yeah I don't know how I reproduced it (whether it was with your attached fg, or a quick FG of my own without graphical elements, but me tagging it "bug" means I saw something.

Could kick my own behind for not writing down what I saw.

**Comment by eltos** [comment]
I don't know if this is what could be happening, but here's a scenario to be investigated:

In principle Rational resampler is implemented to partially produce output. This is useful, especially for large interpolations. Since it calls `set_output_multiple(1)`, it's `general_work` may be given an `noutput_items` that is smaller than `interpolation` * `ninput_items`, say 16 < 25*1. In this case the incoming sample is not consumed because it is still needed in the next call to `general_work` to produce the remaining 9 samples. So far all correct.

However, since the incoming sample is not consumed, this also means that its tag is not yet propagated! Now we are in a situation where the block has produced 16 samples, the first of which should carry the tag from the incoming sample, but it does not yet do so. Now a downstream block could pick up the first of these 16 samples, and process it without knowing about tag it's supposed to carry. When the Resampler block produces the remaining 9 samples (or more) in the next iteration and consumes the input sample, the tag will finally be propagated to a sample, which at this point had already been processed by the downstream block, so the tag is effectively lost!

If this is indeed what's happening here (to be confirmed), the solution would be to change the condition for tag propagation to not only propagate tags associated with just consumed input samples, but in addition also those that — after propagation — are associated with the just produced output samples.

The probability for this unfortunate timing to happen should increase with smaller noutput_items chucks size and with a higher processing rate, both of which are proportional to the sampling rate, so this could explain it barely happens at low rates.

**Comment by eltos** [comment]
Thinking about this a bit more raises another question: in the scheduler, are there any mechanisms that prevent a downstream block from picking up samples during the time between a call to `produce` or `produce_each` from within a blocks `general_work` (when using `WORK_CALLED_PRODUCE`) and the subsequent propagation of tags by the block executor? If not, the tag loss could even happen during a single iteration. Well, not for rational resampler as it does not use `WORK_CALLED_PRODUCE` but rather returns the number of produced samples, and in the block executor the call to `produce_each` for this case happens after tag propagation. But potentially for other blocks using `WORK_CALLED_PRODUCE`? Can someone confirm or disprove?

**Comment by argilo** [comment]
I encountered this bug as well, while attempting to add latency control to gr-rds, in the style of gr-latency_manager. The design does not tolerate dropped tags, but a Rational Resampler block is mysteriously dropping them.

**Comment by chl15819762247-dev** [comment]
Hi 
I'm a beginner in communication engineering and this will be my first open-source contribution.  
I'd like to work on this good first issue by adding the missing docstring explaining the √2 normalization in QPSK soft decision (to keep average symbol energy = 1).  
Thank you!

**Comment by spiccinini** [comment]
Hi! We have also seen this issue in the wild also with the pfb_arb_resampler.  I think that eltos analysis is correct:

> In principle Rational resampler is implemented to partially produce output. This is useful, especially for large interpolations. Since it calls set_output_multiple(1), it's general_work may be given an noutput_items that is smaller than interpolation * ninput_items, say 16 < 25*1. In this case the incoming sample is not consumed because it is still needed in the next call to general_work to produce the remaining 9 samples. So far all correct.

> However, since the incoming sample is not consumed, this also means that its tag is not yet propagated! Now we are in a situation where the block has produced 16 samples, the first of which should carry the tag from the incoming sample, but it does not yet do so. Now a downstream block could pick up the first of these 16 samples, and process it without knowing about tag it's supposed to carry. When the Resampler block produces the remaining 9 samples (or more) in the next iteration and consumes the input sample, the tag will finally be propagated to a sample, which at this point had already been processed by the downstream block, so the tag is effectively lost!

Here is a quick way to reproduce it, that also triggers the same issue with the pfb_arb_resampler.

```python
#!/usr/bin/env python
from gnuradio import gr, gr_unittest, blocks, filter
import pmt

def run(block, N=8000, stride=25, max_nout=None):
    tb = gr.top_block()
    data = [float(i) for i in range(N)]
    tags = []
    for i in range(0, N, stride):
        t = gr.tag_t()
        t.offset = i
        t.key = pmt.intern("test")
        t.value = pmt.from_uint64(i)
        tags.append(t)

    src = blocks.vector_source_f(data, False, 1, tags)
    snk = blocks.vector_sink_f()

    tb.connect(src, block, snk)
    if max_nout is not None:
        block.set_max_noutput_items(max_nout)

    tb.run()

    in_offsets = set(range(0, N, stride))
    out_offsets = set(pmt.to_uint64(t.value) for t in snk.tags())
    return in_offsets, out_offsets


class test_resampler_tag_propagation(gr_unittest.TestCase):
    def test_000_control(self):
        # A rate-1 block must not lose tags under the same constraints.
        blk = blocks.multiply_const_ff(1.0)
        in_offsets, out_offsets = run(blk, max_nout=16)
        self.assertEqual(in_offsets, out_offsets)

    def test_001_rational_resampler_upsample(self):
        blk = filter.rational_resampler_fff(interpolation=25, decimation=7)
        in_offsets, out_offsets = run(blk, N=20000, stride=50, max_nout=16)
        self.assertEqual(in_offsets, out_offsets)

    def test_002_pfb_arb_resampler_upsample(self):
        # output_multiple == filter_size, so the race is timing-dependent
        # Aggregate across multiple rates and chunk sizes
        nfilts = 32
        taps = filter.firdes.low_pass(nfilts, nfilts, 0.4, 0.1)
        blk = filter.pfb_arb_resampler_fff(3.57, taps, nfilts)
        in_offsets, out_offsets = run(blk, N=20000, stride=25, max_nout=96)
        self.assertEqual(in_offsets, out_offsets)

    def test_003_pfb_arb_resampler_downsample(self):
        nfilts = 32
        taps = filter.firdes.low_pass(nfilts, nfilts, 0.4, 0.1)
        blk = filter.pfb_arb_resampler_fff(0.4, taps, nfilts)
        in_offsets, out_offsets = run(blk, N=8000, stride=25, max_nout=32)
        self.assertEqual(in_offsets, out_offsets)


if __name__ == '__main__':
    gr_unittest.run(test_resampler_tag_propagation)
```
